src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -1046,21 +1048,28 @@ private List<Entitlement> findByStackIds(Consumer consumer, List<String> stackId
         if (stackIds == null || stackIds.isEmpty()) {
             return Collections.emptyList();
         }
-        Criteria criteria = currentSession().createCriteria(Entitlement.class)
-            .createAlias("pool", "ent_pool")
-            .createAlias("ent_pool.product", "product")
-            .createAlias("product.attributes", "attrs")
-            .add(Restrictions.eq("attrs.indices", Product.Attributes.STACKING_ID))
-            .add(Restrictions.in("attrs.elements", stackIds))
-            .add(Restrictions.isNull("ent_pool.sourceEntitlement"))
-            .createAlias("ent_pool.sourceStack", "ss", org.hibernate.sql.JoinType.LEFT_OUTER_JOIN)
-            .add(Restrictions.isNull("ss.id"));
+        EntityManager em = this.getEntityManager();
+        CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();
+        CriteriaQuery<Entitlement> criteriaQuery = criteriaBuilder.createQuery(Entitlement.class);
+        Root<Entitlement> entRoot = criteriaQuery.from(Entitlement.class);
+        Join<Entitlement, Pool> poolJoin = entRoot.join(Entitlement_.POOL);
+        Join<Pool, Product> productJoin = poolJoin.join(Pool_.PRODUCT);
+        MapJoin<Product, String, String> attributes = productJoin.joinMap(Product_.ATTRIBUTES);
+        Join<Pool, SourceStack> stackJoin = poolJoin.join(Pool_.SOURCE_STACK, JoinType.LEFT);
+        criteriaQuery.select(entRoot);
+
+        List<Predicate> predicates = new ArrayList<>();
+        predicates.add(criteriaBuilder.equal(attributes.key(), Product.Attributes.STACKING_ID));
+        predicates.add(inPredicate(criteriaBuilder, attributes, stackIds));
+        predicates.add(criteriaBuilder.isNull(poolJoin.get(Pool_.SOURCE_ENTITLEMENT)));
+        predicates.add(criteriaBuilder.isNull(stackJoin.get(SourceStack_.ID)));
 
         if (consumer != null) {
-            criteria.add(Restrictions.eq("consumer", consumer));
+            predicates.add(criteriaBuilder.equal(entRoot.get(Entitlement_.CONSUMER), consumer));
         }
-
-        return (List<Entitlement>) criteria.list();
+        Predicate[] predicateArray = new Predicate[predicates.size()];
+        criteriaQuery.where(predicates.toArray(predicateArray));
+        return em.createQuery(criteriaQuery).getResultList();
```

Comment:
It might be better to convert this to HQL as we don't need to do anything dynamic here.

Same with the `findUpstreamEntitlementForStack`.

Reply:
In one of the CandlepinQuery PR's, the review requested that I use a JPA query instead of HQL when I had a condition for the consumer in the where clause.

https://github.com/candlepin/candlepin/pull/4543

Either is fine by me, but we should at least have a standard.

Reply:
If it's a simple conditional part of WHERE such as `consumer != null` it can be expressed in query. I don't mind using criteria, but I think we should prefer HQL where possible.

Something like the following where the last AND replaces the condition.
```java
    String hql = "SELECT e FROM Entitlement e " +
                 "JOIN e.pool p " +
                 "JOIN p.product pr " +
                 "JOIN pr.attributes a " +
                 "LEFT JOIN p.sourceStack s " +
                 "WHERE a.key = :key AND a.value IN :stackIds " +
                 "AND p.sourceEntitlement IS NULL " +
                 "AND s.id IS NULL " +
                 "AND (e.consumer = :consumer OR :consumer IS NULL)";
```

Reply:
Before we do something like this, I would like to see what the query planner for Postgres and MariaDB/MySQL do with it in the cases where we pass in a null and a non-null. I would hope they're smart enough to short-circuit on that last conditional, but I would like to see verification before we go about throwing our work to the DB in the name of avoiding JPA criteria.

This is one of those cases where standardization is going to be hard. There are conventions and habits that are functional but dangerous for maintenance purposes (string concat), some that are safe but harder to read (JPA criteria), and then we have things like this which are easy to read and avoid concat, but are entirely contingent on how our backend works when it comes to efficiency.

The one thing I want to avoid with any standardization we do is stifle creativity or picking the "correct" or "better" answer because the standard is too strict.

Reply:
As it turns out, consumer_id cannot be null in the entitlement table. The dynamic part is whether we want to filter on consumer or not.

It means the condition of conusmer != null for the query string concat *and* the parameter assignment. Based on that, I'm inclined to keep the CriteriaBuilder code.